Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -160,48 +142,33 @@ class TopSitesPanel: UIViewController {
     }
 
     private func reloadTopSitesWithLimit(limit: Int) -> Success {
-        return self.profile.history.getTopSitesWithLimit(limit).bindQueue(dispatch_get_main_queue()) { result in
-            self.updateDataSourceWithSites(result)
+        return invalidateTopSitesCursor().bindQueue(dispatch_get_main_queue()) { _ in
+            self.dataSource.setDataFromCursor(self.cachedCursor, usingLayout: self.layout)
+            self.dataSource.profile = self.profile
+
+            // redraw now that we've updated our sources
+            self.collection?.collectionViewLayout.invalidateLayout()
+            self.collection?.setNeedsLayout()
             self.collection?.reloadData()
             return succeed()
         }
     }
 
-    private func deleteOrUpdateSites(result: Maybe<Cursor<Site>>, indexPath: NSIndexPath) {
-        guard let collectionView = collection else { return }
-        // get the number of top sites items we have before we update the data sourcce 
-        // this is so we know how many new top sites cells to add
-        // as a sync may have brought in more results than we had previously
-        let previousNumOfThumbnails = collectionView.dataSource?.collectionView(collectionView, numberOfItemsInSection: 0) ?? 0
-
-        // Exit early if the query failed in some way.
-        guard result.isSuccess else {
-            return
+    private func invalidateTopSitesCursor() -> Success {
+        return profile.history.getTopSitesWithLimit(maxFrecencyLimit).bind { result in
+            self.cachedCursor = result.successValue
```

Comment:
You should use `bindQueue` here so you always set the cached cursor on the main thread.